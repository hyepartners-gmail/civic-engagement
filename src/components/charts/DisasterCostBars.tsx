"use client";

import { useMemo } from 'react';
import { BarChart, Bar, XAxis, YAxis, CartesianGrid, Tooltip, Legend, ResponsiveContainer } from 'recharts';

interface DisasterCostBarsProps {
  series: Record<string, [number, number | null][]>;
  stacked: boolean;
  xAxisLabel?: string;
  yAxisLabel?: string;
}

export default function DisasterCostBars({ 
  series, 
  stacked, 
  xAxisLabel = 'Year', 
  yAxisLabel = 'Cost (USD)' 
}: DisasterCostBarsProps) {
  // Convert series to Recharts format
  const chartData = useMemo(() => {
    // Get all years from all series
    const years = new Set<number>();
    Object.values(series).forEach(s => {
      s.forEach(([year]) => years.add(year));
    });
    
    // Sort years
    const sortedYears = Array.from(years).sort((a, b) => a - b);
    
    // Create data points for each year
    return sortedYears.map(year => {
      const dataPoint: Record<string, number | null> = { year };
      Object.entries(series).forEach(([type, s]) => {
        const value = s.find(([y]) => y === year)?.[1] ?? null;
        dataPoint[type] = value;
      });
      return dataPoint;
    });
  }, [series]);
  
  // Get disaster types for coloring
  const disasterTypes = useMemo(() => Object.keys(series), [series]);
  
  // Define colors for different disaster types
  const colors = [
    '#3b82f6', // blue
    '#ef4444', // red
    '#10b981', // green
    '#f59e0b', // amber
    '#8b5cf6', // violet
    '#ec4899', // pink
  ];
  
  if (chartData.length === 0) {
    return (
      <div className="h-80 flex items-center justify-center text-gray-500">
        <p>No disaster cost data available</p>
      </div>
    );
  }

  return (
    <div className="h-80">
      <ResponsiveContainer width="100%" height="100%">
        <BarChart
          data={chartData}
          margin={{ top: 20, right: 30, left: 20, bottom: 60 }}
        >
          <CartesianGrid strokeDasharray="3 3" />
          <XAxis 
            dataKey="year" 
            angle={-45} 
            textAnchor="end" 
            height={60}
            tick={{ fontSize: 12 }}
            label={{ 
              value: xAxisLabel, 
              position: 'insideBottom', 
              offset: -40 
            }}
          />
          <YAxis 
            tick={{ fontSize: 12 }}
            tickFormatter={(value) => {
              if (value >= 1e9) return `$${(value / 1e9).toFixed(1)}B`;
              if (value >= 1e6) return `$${(value / 1e6).toFixed(1)}M`;
              if (value >= 1e3) return `$${(value / 1e3).toFixed(1)}K`;
              return `$${value}`;
            }}
            label={{ 
              value: yAxisLabel, 
              angle: -90, 
              position: 'insideLeft' 
            }}
          />
          <Tooltip 
            formatter={(value) => [`$${Number(value).toLocaleString()}`, 'Cost']}
            labelFormatter={(year) => `Year: ${year}`}
          />
          <Legend />
          {disasterTypes.map((type, index) => (
            <Bar
              key={type}
              dataKey={type}
              stackId={stacked ? "a" : undefined}
              fill={colors[index % colors.length]}
              name={type.charAt(0).toUpperCase() + type.slice(1)}
            />
          ))}
        </BarChart>
      </ResponsiveContainer>
    </div>
  );
}