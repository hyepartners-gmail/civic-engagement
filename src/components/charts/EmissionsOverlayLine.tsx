"use client";

import { useMemo } from 'react';
import { LineChart, Line, XAxis, YAxis, Tooltip, ResponsiveContainer } from 'recharts';

interface EmissionsOverlayLineProps {
  series: [number, number | null][];
  yAxisLabel?: string;
}

export default function EmissionsOverlayLine({ 
  series, 
  yAxisLabel = 'CO₂ Emissions' 
}: EmissionsOverlayLineProps) {
  // Convert series to Recharts format
  const chartData = useMemo(() => {
    return series
      .filter(([_, value]) => value !== null)
      .map(([year, value]) => ({
        year,
        emissions: value
      }));
  }, [series]);
  
  // Determine domain
  const domain = useMemo(() => {
    if (chartData.length === 0) return [0, 100];
    
    const values = chartData.map(d => d.emissions as number);
    const min = Math.min(...values);
    const max = Math.max(...values);
    const range = max - min;
    
    return [min - range * 0.1, max + range * 0.1];
  }, [chartData]);
  
  if (chartData.length === 0) {
    return null; // Don't render if no data
  }

  return (
    <div className="absolute inset-0 pointer-events-none">
      <ResponsiveContainer width="100%" height="100%">
        <LineChart data={chartData}>
          <XAxis 
            dataKey="year" 
            hide={true}
          />
          <YAxis 
            yAxisId="right"
            orientation="right"
            domain={domain}
            hide={true}
          />
          <Tooltip 
            formatter={(value) => [`${Number(value).toLocaleString()} tons`, 'CO₂ Emissions']}
            labelFormatter={(year) => `Year: ${year}`}
          />
          <Line 
            yAxisId="right"
            type="monotone" 
            dataKey="emissions" 
            stroke="#10b981" 
            strokeWidth={2}
            dot={false}
            activeDot={{ r: 4 }}
          />
        </LineChart>
      </ResponsiveContainer>
    </div>
  );
}