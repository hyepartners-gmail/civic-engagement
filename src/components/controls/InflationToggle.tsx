"use client";

import { useUrlState } from '@/hooks/useUrlState';
import { Label } from "@/components/ui/label";
import { Switch } from "@/components/ui/switch";

interface InflationToggleProps {
  enabled: boolean;
  onToggle: (enabled: boolean) => void;
}

export default function InflationToggle({ enabled, onToggle }: InflationToggleProps) {
  const [urlReal, setUrlReal] = useUrlState<boolean>('real', true);

  const handleToggle = (newState: boolean) => {
    onToggle(newState);
    setUrlReal(newState);
  };

  return (
    <div className="flex items-center space-x-2">
      <Label htmlFor="inflation-toggle">Adjust for Inflation</Label>
      <Switch
        id="inflation-toggle"
        checked={enabled}
        onCheckedChange={handleToggle}
      />
    </div>
  );
}