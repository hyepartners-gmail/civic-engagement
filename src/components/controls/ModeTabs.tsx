"use client";

import { useUrlState } from '@/hooks/useUrlState';
import { Tabs, TabsList, TabsTrigger } from "@/components/ui/tabs";

interface ModeTabsProps {
  mode: 'total' | 'percap';
  onModeChange: (mode: 'total' | 'percap') => void;
}

export default function ModeTabs({ mode, onModeChange }: ModeTabsProps) {
  const [urlMode, setUrlMode] = useUrlState<'total' | 'percap'>('costMode', 'total');

  const handleModeChange = (newMode: string) => {
    const mode = newMode as 'total' | 'percap';
    onModeChange(mode);
    setUrlMode(mode);
  };

  return (
    <Tabs value={mode} onValueChange={handleModeChange}>
      <TabsList>
        <TabsTrigger value="total">Total $</TabsTrigger>
        <TabsTrigger value="percap">Per-capita $</TabsTrigger>
      </TabsList>
    </Tabs>
  );
}