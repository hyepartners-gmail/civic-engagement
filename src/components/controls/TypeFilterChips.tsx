"use client";

import { useState, useEffect } from 'react';
import { useUrlState } from '@/hooks/useUrlState';

interface TypeFilterChipsProps {
  types: string[];
  allTypes: string[];
}

export default function TypeFilterChips({ types, allTypes }: TypeFilterChipsProps) {
  const [encodedTypes, setEncodedTypes] = useUrlState<string | null>('types', null);
  
  // Update URL when types change
  useEffect(() => {
    if (types.length === 0 || types.length === allTypes.length) {
      setEncodedTypes(null); // Don't encode if all types are selected
    } else {
      setEncodedTypes(types.join('|'));
    }
  }, [types, allTypes.length, setEncodedTypes]);
  
  const toggleType = (type: string) => {
    if (types.includes(type)) {
      // Remove type
      const newTypes = types.filter(t => t !== type);
      // If removing the last type, select all types
      setEncodedTypes(newTypes.length > 0 ? newTypes.join('|') : null);
    } else {
      // Add type
      const newTypes = [...types, type];
      setEncodedTypes(newTypes.join('|'));
    }
  };
  
  const selectAll = () => {
    setEncodedTypes(null); // Reset to default (all types)
  };
  
  const clearAll = () => {
    // This would effectively select no types, which isn't useful
    // Instead, we'll just leave it as is
  };

  return (
    <div className="flex flex-wrap gap-2">
      {allTypes.map(type => {
        const isSelected = types.includes(type);
        const displayName = type.charAt(0).toUpperCase() + type.slice(1).replace(/([A-Z])/g, ' $1');
        
        return (
          <button
            key={type}
            onClick={() => toggleType(type)}
            className={`px-3 py-1 rounded-full text-sm font-medium transition-colors ${
              isSelected
                ? 'bg-blue-500 text-white'
                : 'bg-gray-200 text-gray-700 hover:bg-gray-300'
            }`}
            aria-pressed={isSelected}
          >
            {displayName}
          </button>
        );
      })}
      
      <button
        onClick={selectAll}
        className="px-3 py-1 rounded-full text-sm font-medium bg-gray-100 text-gray-700 hover:bg-gray-200"
      >
        Select All
      </button>
    </div>
  );
}