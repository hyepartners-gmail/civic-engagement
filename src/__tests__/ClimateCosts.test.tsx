import { render } from '@testing-library/react';
import DisasterCostBars from '@/components/charts/DisasterCostBars';
import EmissionsOverlayLine from '@/components/charts/EmissionsOverlayLine';
import TypeFilterChips from '@/components/controls/TypeFilterChips';
import ModeTabs from '@/components/controls/ModeTabs';
import InflationToggle from '@/components/controls/InflationToggle';

// Mock data for testing
const mockFemaCosts = {
  hurricane: [
    [2000, 1000000],
    [2001, 1200000],
    [2002, 1100000]
  ],
  flood: [
    [2000, 800000],
    [2001, 900000],
    [2002, 850000]
  ]
};

const mockEmissions = [
  [2000, 5500],
  [2001, 5400],
  [2002, 5300]
];

const mockAllTypes = ['hurricane', 'flood', 'wildfire', 'drought', 'severeStorm'];

describe('Climate Costs Components', () => {
  describe('DisasterCostBars', () => {
    it('should render stacked bars correctly', () => {
      const { container } = render(
        <DisasterCostBars 
          series={mockFemaCosts} 
          stacked={true}
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render grouped bars correctly', () => {
      const { container } = render(
        <DisasterCostBars 
          series={mockFemaCosts} 
          stacked={false}
        />
      );
      
      expect(container).toMatchSnapshot();
    });
  });

  describe('EmissionsOverlayLine', () => {
    it('should render correctly with data', () => {
      const { container } = render(
        <EmissionsOverlayLine 
          series={mockEmissions} 
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render nothing when no data', () => {
      const { container } = render(
        <EmissionsOverlayLine 
          series={[]} 
        />
      );
      
      expect(container).toMatchSnapshot();
    });
  });

  describe('TypeFilterChips', () => {
    it('should render correctly with all types selected', () => {
      const { container } = render(
        <TypeFilterChips 
          types={mockAllTypes} 
          allTypes={mockAllTypes} 
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render correctly with some types selected', () => {
      const { container } = render(
        <TypeFilterChips 
          types={['hurricane', 'flood']} 
          allTypes={mockAllTypes} 
        />
      );
      
      expect(container).toMatchSnapshot();
    });
  });

  describe('ModeTabs', () => {
    it('should render correctly in total mode', () => {
      const { container } = render(
        <ModeTabs 
          mode="total" 
          onModeChange={jest.fn()} 
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render correctly in percap mode', () => {
      const { container } = render(
        <ModeTabs 
          mode="percap" 
          onModeChange={jest.fn()} 
        />
      );
      
      expect(container).toMatchSnapshot();
    });
  });

  describe('InflationToggle', () => {
    it('should render correctly when enabled', () => {
      const { container } = render(
        <InflationToggle 
          enabled={true} 
          onToggle={jest.fn()} 
        />
      );
      
      expect(container).toMatchSnapshot();
    });

    it('should render correctly when disabled', () => {
      const { container } = render(
        <InflationToggle 
          enabled={false} 
          onToggle={jest.fn()} 
        />
      );
      
      expect(container).toMatchSnapshot();
    });
  });
});