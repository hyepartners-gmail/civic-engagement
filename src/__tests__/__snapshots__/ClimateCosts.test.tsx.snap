// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Climate Costs Components EmissionsOverlayLine should render nothing when no data 1`] = `<div />`;
